### PR TITLE
Add Amazon VPC CNI support

### PR DESCRIPTION
**Pull request by Sn0rt**
this PR for Amazone vpc cni support wtih v1alpha2 , v1alpha1 here is #1062  

# how to use

this is different from HEAD of repo, the user should set a bash env variable as follow 

```bash
$ export AWS_REGION=us-east-2
$ export AWS_CREDENTIALS=$(cat ~/aws_cluster_deploy_env.sh)
$ export AWS_ACCESS_KEY_ID=$(echo $AWS_CREDENTIALS | jq .AccessKey.AccessKeyId -r)
$ export AWS_SECRET_ACCESS_KEY=$(echo $AWS_CREDENTIALS | jq .AccessKey.SecretAccessKey -r)
$ export CLUSTER_NAME=sn0rttest
$ export NETWORK=amazon-vpc-cni-k8s
$ rm -rf examples/_out
$ ./examples/generate.sh
Generated /Volumes/home/guohao/workspace/go/src/sigs.k8s.io/cluster-api-provider-aws/examples/_out/addons.yaml
Generated /Volumes/home/guohao/workspace/go/src/sigs.k8s.io/cluster-api-provider-aws/examples/_out/cluster.yaml
Generated /Volumes/home/guohao/workspace/go/src/sigs.k8s.io/cluster-api-provider-aws/examples/_out/controlplane.yaml
Generated /Volumes/home/guohao/workspace/go/src/sigs.k8s.io/cluster-api-provider-aws/examples/_out/machinedeployment.yaml
```

# cluster and machine spec

the cluster file example

```yaml
apiVersion: cluster.x-k8s.io/v1alpha2
kind: Cluster
metadata:
  annotations:
    cluster.k8s.io/network-cni: amazon-vpc-cni-k8s
  name: test1
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - 10.0.0.0/16
    services:
      cidrBlocks:
      - 192.168.0.0/16
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    kind: AWSCluster
    name: test1
    namespace: default
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: AWSCluster
metadata:
  name: test1
  namespace: default
spec:
  region: us-east-2
  sshKeyName: default
```

the KubeadmConfig of test1-controlplane-0
 
```yaml
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
kind: KubeadmConfig
metadata:
  name: test1-controlplane-0
  namespace: default
spec:
  clusterConfiguration:
    apiServer:
      extraArgs:
        cloud-provider: aws
    controllerManager:
      extraArgs:
        cloud-provider: aws
  initConfiguration:
    nodeRegistration:
      kubeletExtraArgs:
        cloud-provider: aws
        max-pods: 19
        node-ip: '{{ ds.meta_data.hostname }}'
      name: '{{ ds.meta_data.hostname }}'
```

the AWSMachine of test1-controlplane-0

```
apiVersion: cluster.x-k8s.io/v1alpha2
kind: Machine
metadata:
  labels:
    cluster.x-k8s.io/cluster-name: test1
    cluster.x-k8s.io/control-plane: "true"
  name: test1-controlplane-0
spec:
  bootstrap:
    configRef:
      apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
      kind: KubeadmConfig
      name: test1-controlplane-0
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    kind: AWSMachine
    name: test1-controlplane-0
    namespace: default
  version: v1.15.3
```

the AWSMachine of test1-controlplane-0

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: AWSMachine
metadata:
  name: test1-controlplane-0
  namespace: default
spec:
  iamInstanceProfile: controllers.cluster-api-provider-aws.sigs.k8s.io
  instanceType: t2.medium
  sshKeyName: default
```

the machine deployment

```yaml
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
kind: KubeadmConfigTemplate
metadata:
  name: test1-md-0
  namespace: default
spec:
  template:
    spec:
      joinConfiguration:
        nodeRegistration:
          kubeletExtraArgs:
            cloud-provider: aws
            max-pods: 19
            node-ip: '{{ ds.meta_data.hostname }}'
          name: '{{ ds.meta_data.hostname }}'
```

# security group info

Drawing on the security group policy of EKS, nodes in the cluster directly communicate with each other by default.

### CODE REVIEW

File: pkg/cloud/services/ec2/securitygroups.go [hunk 1]
```diff
@@ -47,6 +46,15 @@ const (
 	IPProtocolICMPv6 = "58"
 )
 
+const (
+	// user set specify cni
+	clusterSetCNI = "cluster.k8s.io/network-cni"
+
+	// the amazon native network solution
+	// https://github.com/aws/amazon-vpc-cni-k8s
+	amazonVPCCNI = "amazon-vpc-cni-k8s"
```
Comment (randomvariable) [inline pkg/cloud/services/ec2/securitygroups.go]: Would like to follow K8s convention for CamelCase on field values and use `AmazonVPC`

File: pkg/cloud/services/cloudformation/bootstrap.go [hunk 2]
```diff
@@ -296,6 +296,28 @@ func cloudProviderControlPlaneAwsPolicy() *iam.PolicyDocument {
...
+					"ec2:UnassignPrivateIpAddresses",
+				},
+			},
+			{
+				Effect:   iam.EffectAllow,
+				Resource: iam.Resources{"arn:aws:ec2:*:*:network-interface/*"},
+				Action: iam.Actions{
+					"ec2:CreateTags",
+				},
+			},
```
Comment (aaroniscode) [inline pkg/cloud/services/cloudformation/bootstrap.go]: Would it make sense to make optional?  Amazon best practices is to grant [Least Privilege](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#grant-least-privilege).

The question is how to make this optional? Would this be best exposed in `clusterawsadm` as a flag?

Maybe `--cni amazon-vpc-cni`.  Not sure there are other CNI's that would require different policy.

or `--add-vpc-cni-policy`

Thoughts on if this should be optional and if so, how best to expose?

File: pkg/cloud/services/ec2/securitygroups.go (hunk 1 above)
Comment (Sn0rt) [reply pkg/cloud/services/ec2/securitygroups.go]: > Would like to follow K8s convention for CamelCase on field values and use `AmazonVPC`?

copy that.

File: pkg/cloud/services/cloudformation/bootstrap.go (hunk 2 above)
Comment (Sn0rt) [reply pkg/cloud/services/cloudformation/bootstrap.go]: someone wants to create mutil clusters with different network solution, such as calico and the amazon-vpc-cni, under one AWS account. 
that use case how to separate the privilege of account ?

Comment (randomvariable) [reply pkg/cloud/services/cloudformation/bootstrap.go]: Make it accept multi-value with comma-separation?

Options:
- Calico
- AmazonVPC

Default: [Calico]

Comment (Sn0rt) [reply pkg/cloud/services/cloudformation/bootstrap.go]: agree with you, we can talk about this issues later.

**Comment by Sn0rt** [comment]
cluster yaml

```yaml
apiVersion: cluster.x-k8s.io/v1alpha2
kind: Cluster
metadata:
  annotations:
    cluster.k8s.io/network-cni: AmazonVPC
  name: newcluster
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - 10.0.0.0/16
    services:
      cidrBlocks:
      - 192.168.0.0/16
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    kind: AWSCluster
    name: newcluster
    namespace: default
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: AWSCluster
metadata:
  name: newcluster
  namespace: default
spec:
  region: us-east-2
  sshKeyName: guohao
```

node info
```
$ kubectl --kubeconfig guohao4 get node -o wide --all-namespaces
NAME                                       STATUS   ROLES    AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION    CONTAINER-RUNTIME
ip-10-0-0-195.us-east-2.compute.internal   Ready    master   5m39s   v1.15.3   10.0.0.195    <none>        Ubuntu 18.04.3 LTS   4.15.0-1045-aws   containerd://1.2.8
ip-10-0-0-196.us-east-2.compute.internal   Ready    master   5m37s   v1.15.3   10.0.0.196    <none>        Ubuntu 18.04.3 LTS   4.15.0-1045-aws   containerd://1.2.8
ip-10-0-0-208.us-east-2.compute.internal   Ready    master   6m44s   v1.15.3   10.0.0.208    <none>        Ubuntu 18.04.3 LTS   4.15.0-1045-aws   containerd://1.2.8
ip-10-0-0-39.us-east-2.compute.internal    Ready    <none>   62s     v1.15.3   10.0.0.39     <none>        Ubuntu 18.04.3 LTS   4.15.0-1045-aws   containerd://1.2.8
```

pod info
```
$ kubectl --kubeconfig guohao4 get pod -o wide --all-namespaces
NAMESPACE     NAME                                                               READY   STATUS    RESTARTS   AGE     IP           NODE                                       NOMINATED NODE   READINESS GATES
2048-game     2048-deployment-c99556b4b-sr99d                                    1/1     Running   0          2m56s   10.0.0.189   ip-10-0-0-39.us-east-2.compute.internal    <none>           <none>
2048-game     2048-deployment-c99556b4b-vmwzm                                    1/1     Running   0          2m56s   10.0.0.92    ip-10-0-0-39.us-east-2.compute.internal    <none>           <none>
kube-system   aws-node-9qm7j                                                     1/1     Running   0          5m16s   10.0.0.208   ip-10-0-0-208.us-east-2.compute.internal   <none>           <none>
kube-system   aws-node-m4sgr                                                     1/1     Running   0          73s     10.0.0.39    ip-10-0-0-39.us-east-2.compute.internal    <none>           <none>
kube-system   aws-node-mcb5r                                                     1/1     Running   0          5m16s   10.0.0.195   ip-10-0-0-195.us-east-2.compute.internal   <none>           <none>
kube-system   aws-node-n28js                                                     1/1     Running   0          5m16s   10.0.0.196   ip-10-0-0-196.us-east-2.compute.internal   <none>           <none>
kube-system   coredns-5c98db65d4-gwhjs                                           1/1     Running   0          6m37s   10.0.0.174   ip-10-0-0-196.us-east-2.compute.internal   <none>           <none>
kube-system   coredns-5c98db65d4-xf52h                                           1/1     Running   0          6m37s   10.0.0.64    ip-10-0-0-196.us-east-2.compute.internal   <none>           <none>
kube-system   etcd-ip-10-0-0-195.us-east-2.compute.internal                      1/1     Running   0          5m50s   10.0.0.195   ip-10-0-0-195.us-east-2.compute.internal   <none>           <none>
kube-system   etcd-ip-10-0-0-196.us-east-2.compute.internal                      1/1     Running   0          5m35s   10.0.0.196   ip-10-0-0-196.us-east-2.compute.internal   <none>           <none>
kube-system   etcd-ip-10-0-0-208.us-east-2.compute.internal                      1/1     Running   0          6m2s    10.0.0.208   ip-10-0-0-208.us-east-2.compute.internal   <none>           <none>
kube-system   kube-apiserver-ip-10-0-0-195.us-east-2.compute.internal            1/1     Running   0          4m57s   10.0.0.195   ip-10-0-0-195.us-east-2.compute.internal   <none>           <none>
kube-system   kube-apiserver-ip-10-0-0-196.us-east-2.compute.internal            1/1     Running   1          5m33s   10.0.0.196   ip-10-0-0-196.us-east-2.compute.internal   <none>           <none>
kube-system   kube-apiserver-ip-10-0-0-208.us-east-2.compute.internal            1/1     Running   0          5m34s   10.0.0.208   ip-10-0-0-208.us-east-2.compute.internal   <none>           <none>
kube-system   kube-controller-manager-ip-10-0-0-195.us-east-2.compute.internal   1/1     Running   0          4m32s   10.0.0.195   ip-10-0-0-195.us-east-2.compute.internal   <none>           <none>
kube-system   kube-controller-manager-ip-10-0-0-196.us-east-2.compute.internal   1/1     Running   0          4m47s   10.0.0.196   ip-10-0-0-196.us-east-2.compute.internal   <none>           <none>
kube-system   kube-controller-manager-ip-10-0-0-208.us-east-2.compute.internal   1/1     Running   1          5m57s   10.0.0.208   ip-10-0-0-208.us-east-2.compute.internal   <none>           <none>
kube-system   kube-proxy-72mpq                                                   1/1     Running   0          5m42s   10.0.0.196   ip-10-0-0-196.us-east-2.compute.internal   <none>           <none>
kube-system   kube-proxy-q8vlv                                                   1/1     Running   0          5m50s   10.0.0.195   ip-10-0-0-195.us-east-2.compute.internal   <none>           <none>
kube-system   kube-proxy-svtfz                                                   1/1     Running   0          6m37s   10.0.0.208   ip-10-0-0-208.us-east-2.compute.internal   <none>           <none>
kube-system   kube-proxy-xwglg                                                   1/1     Running   0          73s     10.0.0.39    ip-10-0-0-39.us-east-2.compute.internal    <none>           <none>
kube-system   kube-scheduler-ip-10-0-0-195.us-east-2.compute.internal            1/1     Running   0          4m33s   10.0.0.195   ip-10-0-0-195.us-east-2.compute.internal   <none>           <none>
kube-system   kube-scheduler-ip-10-0-0-196.us-east-2.compute.internal            1/1     Running   0          4m33s   10.0.0.196   ip-10-0-0-196.us-east-2.compute.internal   <none>           <none>
kube-system   kube-scheduler-ip-10-0-0-208.us-east-2.compute.internal            1/1     Running   1          5m58s   10.0.0.208   ip-10-0-0-208.us-east-2.compute.internal   <none>           <none>
```

**Comment by randomvariable** [comment]
Make sure to put the license headers on the new files.

**Comment by ncdc** [comment]
/hold
We have a few small PRs going in to master for a new v0.4.x release before we fully open master for breaking changes / new features for v0.5.0. We'll unhold once that's done. Thanks for your patience!

File: examples/cluster/cluster.yaml [hunk 3]
```diff
@@ -2,11 +2,17 @@
 apiVersion: cluster.x-k8s.io/v1alpha2
 kind: Cluster
 metadata:
+  annotations:
+    cluster.k8s.io/network-cni: ${NETWORK}
```
Comment (randomvariable) [inline examples/cluster/cluster.yaml]:
```suggestion
    cluster.x-k8s.io/network-cni: ${NETWORK}
```

Comment (randomvariable) [reply examples/cluster/cluster.yaml]: We need to adopt the alpha `x-k8s.io` annotations for now, but otherwise good to go.

Comment (Sn0rt) [reply examples/cluster/cluster.yaml]: > We need to adopt the alpha `x-k8s.io` annotations for now, but otherwise good to go.

thx very much.

**Comment by randomvariable** [comment]
Apologies @Sn0rt for leaving this so long.
As much as we would like to enable the choice of the VPC CNI, I don't think this is the right place for it. I would like to revisit this after we have clusterctl v2 in place which is happening as part of v1alpha3. I have created #1484 as an umbrella issue and would welcome your comments.

/close

**Pull request closed without merging**